docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+* Often operators have customizable parameters such as information about locations of external systems or parameters that modify the behavior of the operator. Users should be able to specify these easily without having to change source code. This can be done by making them properties of the operator because they can then be initialized from external properties files.
+	* Where possible default values should be provided for the properties in the source code.
+	* Validation rules should be specified for the properties using javax constraint validations that check whether the values specified for the properties are in the correct format, range or other operator requirements. Required properties should have at least a @NotNull validation specifying that they have to be specified by the user.
+
+### Checkpointing
+
+Checkpointing is a process of snapshotting the state of an operator and saving it so that in case of failure the state can be used to restore the operator and continue processing. It is automatically performed by the platform at a configurable interval. All operators in the application are checkpointed in a distributed fashion, thus allowing the entire state of the application to be saved and available for recovery if needed. Here are some things to remember when it comes to checkpointing
+
+* The process of checkpointing involves snapshotting the state by serializing the operator and saving it to a store. This is done using a **StorageAgent**. By default a *StorageAgent* is already provided by the platform and it is called **AsyncFSStorageAgent**. It serializes the operator using Kryo and saves the serialized state asynchronously to a filesystem such as HDFS. There are other implementations of *StorageAgent* available such as **GeodeKeyValueStorageAgent** that stores the serialized state in Geode which is an in-memory replicated data grid.
+* All variables in the operator not marked transient and final are saved so any variables in the operator that are not part of the state should be marked transient. Specifically any variables like connection objects, i/o streams, ports are transient, because they need to be setup again on failure recovery.
+* If the operator does not keep any state between windows, mark it with the **@Stateless** annotation. This results in efficiencies during checkpointing and recovery. The operator will not be checkpointed and always restored to the initial state
+* The checkpoint interval can be set using the **CHECKPOINT\_WINDOW\_COUNT** attribute which specifies the interval in terms of number of streaming windows.
```

Comment:
Not clear if underscores need to be escaped (the one in SPIN_MILLIS above is not)